### PR TITLE
[CARBONDATA-1827] Added S3 Implementation 

### PR DESCRIPTION
**Pull request by SangeetaGulia**
Be sure to do all of the following checklist to help us incorporate 
your contribution quickly and easily:

 - [ ] Any interfaces changed? 
 No
 - [ ] Any backward compatibility impacted?
 No
 - [ ] Document update required?
No
 - [x] Testing done
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required?
Added new Unit Test Cases.

        - How it is tested? Please attach test report.
        - Is it a performance related change? Please attach the performance test report.
        - Any additional information to help reviewers in testing this change.
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Can one of the admins verify this patch?

**Comment by ravipesala** [comment]
Can one of the admins verify this patch?

**Comment by CarbonDataQA** [comment]
Can one of the admins verify this patch?

File: core/src/main/java/org/apache/carbondata/core/constants/CarbonCommonConstants.java [hunk 1]
```diff
@@ -162,6 +161,12 @@
 
   public static final String S3A_PREFIX = "s3a://";
 
+
+  /**
+   * VIEWS3URL_PREFIX
+   */
+  public static final String S3URL_PREFIX = "s3a://";
```
Comment (mohammadshahidkhan) [inline core/src/main/java/org/apache/carbondata/core/constants/CarbonCommonConstants.java]: S3A_PREFIX  and S3URL_PREFIX both have same constant.
Please remove once of them.

File: core/src/main/java/org/apache/carbondata/core/constants/CarbonCommonConstants.java [hunk 2]
```diff
@@ -1356,6 +1367,35 @@
 
   public static final String CARBON_UPDATE_SYNC_FOLDER_DEFAULT = "/tmp/carbondata";
 
+  /**
+   * S3 Constants
+   */
+
+  public static final String S3_ACCESS_KEY = "fs.s3a.access.key";
```
Comment (mohammadshahidkhan) [inline core/src/main/java/org/apache/carbondata/core/constants/CarbonCommonConstants.java]: i think its better not to keep the s3 specific property is CarbonCommonConstant
fs.s3a.access.key, fs.s3a.secret.key, fs.s3a.impl
these constant already would have been defined in s3 .

File: core/src/main/java/org/apache/carbondata/core/datastore/impl/FileFactory.java [hunk 3]
```diff
@@ -81,24 +77,22 @@ public static FileHolder getFileHolder(FileType fileType) {
...
         return new DFSFileHolderImpl();
+      case S3:
+        return new S3FileHolderImpl();
       default:
         return new FileHolderImpl();
     }
   }
 
   public static FileType getFileType(String path) {
-    String lowerPath = path.toLowerCase();
```
Comment (mohammadshahidkhan) [inline core/src/main/java/org/apache/carbondata/core/datastore/impl/FileFactory.java]: Any reason why converting to lower case is removed.

File: core/src/main/java/org/apache/carbondata/core/locks/CarbonLockFactory.java [hunk 4]
```diff
@@ -52,23 +52,23 @@
    */
   public static ICarbonLock getCarbonLockObj(AbsoluteTableIdentifier absoluteTableIdentifier,
       String lockFile) {
-    switch (lockTypeConfigured) {
-      case CarbonCommonConstants.CARBON_LOCK_TYPE_LOCAL:
```
Comment (mohammadshahidkhan) [inline core/src/main/java/org/apache/carbondata/core/locks/CarbonLockFactory.java]: Why swith case change to if else

Comment (SangeetaGulia) [reply core/src/main/java/org/apache/carbondata/core/locks/CarbonLockFactory.java]: This change is to support multiple filesystems on a single carbonsession. So at a time if you work on S3 and local together, it should implicitly understand what lock type it should provide. 

So we are identifying lock type based on the type of file and the conditional match falls on two variables:  one is lockTypeConfigured and second, we have to identify what type of file it is?(whether s3, hdfs, local) on the basis of which we will set lockTypeConfigured.

Due to this change now we don't need to specify LOCK_TYPE in carbonProperties explicitly for s3, hdfs and local.

**Comment by SangeetaGulia** [comment]
retest this please

**Comment by kunal642** [comment]
retest this please

**Comment by chenliang613** [comment]
add to whitelist

**Comment by SangeetaGulia** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1610/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1613/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2004/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2007/

**Comment by SangeetaGulia** [comment]
@mohammadshahidkhan Can you please review, we have made the changes.

File: core/pom.xml [hunk 5]
```diff
@@ -94,6 +99,20 @@
         </exclusion>
       </exclusions>
     </dependency>
+
+    <!-- https://mvnrepository.com/artifact/org.apache.hadoop/hadoop-aws -->
+    <dependency>
+      <groupId>org.apache.hadoop</groupId>
```
Comment (jackylk) [inline core/pom.xml]: Why is this required? Core should not dependent on hadoop, right?

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/952/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2181/

File: core/src/main/java/org/apache/carbondata/core/datastore/filesystem/S3CarbonFile.java [hunk 6]
```diff
@@ -0,0 +1,445 @@
...
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.fs.permission.FsPermission;
+import org.apache.hadoop.io.compress.BZip2Codec;
+import org.apache.hadoop.io.compress.CompressionCodec;
+import org.apache.hadoop.io.compress.CompressionCodecFactory;
+import org.apache.hadoop.io.compress.GzipCodec;
+
+public class S3CarbonFile implements CarbonFile {
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/datastore/filesystem/S3CarbonFile.java]: Why we need this file, can't we use HDFSCarbonFile directly?

File: core/src/main/java/org/apache/carbondata/core/datastore/filesystem/S3CarbonFile.java [hunk 7]
```diff
@@ -0,0 +1,445 @@
...
+
+  private static final LogService LOGGER =
+      LogServiceFactory.getLogService(S3CarbonFile.class.getName());
+  private static Configuration configuration = null;
+
+  static {
+    configuration = new Configuration();
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/datastore/filesystem/S3CarbonFile.java]: I think we should use the Configuration object from the job instead of creating it ourselves

Comment (QiangCai) [inline core/src/main/java/org/apache/carbondata/core/datastore/filesystem/S3CarbonFile.java]: Carbon should support to get configuration from user session (like SparkSession) to support multiple users.
Different users can use different S3 credential in the same cluster.

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2396/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1173/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2401/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1178/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2406/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1182/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2597/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2599/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2601/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2603/

**Comment by jatin9896** [comment]
retest this please

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2631/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2453/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1232/

File: core/src/main/java/org/apache/carbondata/core/constants/CarbonCommonConstants.java [hunk 8]
```diff
@@ -1420,6 +1425,16 @@
 
   public static final String CARBON_UPDATE_SYNC_FOLDER_DEFAULT = "/tmp/carbondata";
 
+  /**
+   * S3 Constants
+   */
+
+  public static final String S3_IMPLEMENTATION = "fs.s3a.impl";
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/constants/CarbonCommonConstants.java]: How about S3N support? I think this can be a property set by user, user can choose S3 implementation class. The default class can be S3A

File: core/src/main/java/org/apache/carbondata/core/datastore/impl/FileFactory.java [hunk 9]
```diff
@@ -220,7 +221,16 @@ public static boolean mkdirs(String filePath, FileType fileType) throws IOExcept
    */
   public static DataOutputStream getDataOutputStreamUsingAppend(String path, FileType fileType)
       throws IOException {
-    return getCarbonFile(path).getDataOutputStreamUsingAppend(path, fileType);
+    if(FileType.S3 == fileType) {
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/datastore/impl/FileFactory.java]: add space after `if`

File: core/pom.xml [hunk 10]
```diff
@@ -94,6 +99,15 @@
         </exclusion>
       </exclusions>
     </dependency>
+
+
```
Comment (jackylk) [inline core/pom.xml]: remove empty line

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+import org.apache.spark.sql.SparkSession
+
+import org.apache.carbondata.core.constants.CarbonCommonConstants
+import org.apache.carbondata.core.util.CarbonProperties
+
+object S3Example {
+
+  def main(args: Array[String]) {
+
+
```
Comment (jackylk) [inline examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala]: remove empty line

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+    val rootPath = new File(this.getClass.getResource("/").getPath
+                            + "../../../..").getCanonicalPath
+    val storeLocation = s"$rootPath/examples/spark2/target/store"
+    val warehouse = s"$rootPath/examples/spark2/target/warehouse"
+    val metastoredb = s"$rootPath/examples/spark2/target"
+    val path = s"$rootPath/examples/spark2/src/main/resources/data.csv"
+
+
```
Comment (jackylk) [inline examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala]: remove empty line

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+      .addProperty(CarbonCommonConstants.ENABLE_UNSAFE_COLUMN_PAGE_LOADING, "true")
+
+    import org.apache.spark.sql.CarbonSession._
+    val spark = SparkSession
+      .builder()
+      .master("local")
+      .appName("CarbonSessionExample")
+      .config("spark.sql.warehouse.dir", warehouse)
+      .config("spark.driver.host", "localhost")
+      .config(ACCESS_KEY, "*****************")
```
Comment (jackylk) [inline examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala]: it is better to get all S3 related config from command line argument (`args`), and print the usage of this class, like
```
Usage: java CarbonS3Example <fs.s3a.endpoint> <fs.s3a.access.key>" +
                       " <fs.s3a.secret.key> <fs.s3a.impl> <carbon store location>
```

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/FederationExample.scala [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+
+import java.io.File
+
+import org.apache.hadoop.fs.s3a.Constants.{ACCESS_KEY, SECRET_KEY}
+import org.apache.spark.sql.SparkSession
+
+import org.apache.carbondata.core.constants.CarbonCommonConstants
+import org.apache.carbondata.core.util.CarbonProperties
+
+object FederationExample {
```
Comment (jackylk) [inline examples/spark2/src/main/scala/org/apache/carbondata/examples/FederationExample.scala]: What is this example? can you add comment. Add for next example also

File: core/src/main/java/org/apache/carbondata/core/datastore/impl/FileFactory.java [hunk 15]
```diff
@@ -171,6 +171,7 @@ public static boolean createNewFile(
       final FsPermission permission) throws IOException {
     return getCarbonFile(filePath).createNewFile(filePath, fileType, doAs, permission);
   }
+
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/datastore/impl/FileFactory.java]: How about append support? please add subtasks to CARBONDATA-1827:
1. support S3 table with dictionary
2. support compaction on S3 table
3. support data update/delete on S3 table
4. support alter table add columns/drop columns on S3 table

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+      .config(SECRET_KEY, "***************************")
+      .getOrCreateCarbonSession(storeLocation, warehouse)
+
+    spark.sparkContext.setLogLevel("INFO")
+    spark.sql("CREATE DATABASE if not exists s3carbondemo LOCATION 's3a://<bucketName>/s3carbondemo'")
```
Comment (jackylk) [inline examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala]: add a DROP TABLE before creating it

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+         | complexData ARRAY<STRING>
+         | )
+         | STORED BY 'carbondata'
+         | TBLPROPERTIES('SORT_COLUMNS'='', 'DICTIONARY_INCLUDE'='dateField, charField')
+       """.stripMargin)
+
+    // scalastyle:off
+    spark.sql(
+      s"""
+         | LOAD DATA LOCAL INPATH '$path'
```
Comment (jackylk) [inline examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala]: add another load, so that it will become two segments

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2640/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1237/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2461/

**Comment by jackylk** [comment]
style check error for newly added examples

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1345/

**Comment by jatin9896** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2590/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2753/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1355/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2595/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1361/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1391/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2626/

**Comment by jatin9896** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2668/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1434/

**Comment by SangeetaGulia** [comment]
@jackylk we have made the changes as per your review comments. Also, CI build is failing due to some Jenkins issue : 

Build finished. 'FileNotFoundException means that the credentials Jenkins is using are probably wrong. Or the user account does not have write access to the repo.

Can you please check.

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2675/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1441/

**Comment by SangeetaGulia** [comment]
@jackylk we have made the changes as per your review comments. Can you please check.

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2813/

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala [hunk 18]
```diff
@@ -0,0 +1,152 @@
...
+         | bigintField LONG,
+         | doubleField DOUBLE,
+         | stringField STRING,
+         | timestampField TIMESTAMP,
+         | decimalField DECIMAL(18,2),
+         | dateField DATE,
+         | charField CHAR(5),
+         | floatField FLOAT
+         | )
+         | STORED BY 'carbondata'
```
Comment (jackylk) [inline examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala]: I think this example should use `CREATE TABLE ... LOCATION 's3a://from_cmd_param'` to create table on s3.

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala [hunk 19]
```diff
@@ -0,0 +1,152 @@
...
+    val (accessKey, secretKey) = getKeyOnPrefix(args(2))
+    val spark = SparkSession
+      .builder()
+      .master("local")
+      .appName("CarbonSessionExample")
+      .config("spark.sql.warehouse.dir", warehouse)
+      .config("spark.driver.host", "localhost")
+      .config(accessKey, args(0))
+      .config(secretKey, args(1))
+      .getOrCreateCarbonSession(args(2), warehouse)
```
Comment (jackylk) [inline examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala]: Do not use store path, now carbon supports specifying location when creating table. This is the standard way to create table. So user can create one session and decide the table storage location in CREATE TABLE DDL, it can be HDFS, S3, or others.

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala [hunk 20]
```diff
@@ -0,0 +1,152 @@
...
+   */
+
+  def main(args: Array[String]) {
+    val rootPath = new File(this.getClass.getResource("/").getPath
+                            + "../../../..").getCanonicalPath
+    val warehouse = s"$rootPath/examples/spark2/target/warehouse"
+    val path = s"$rootPath/examples/spark2/src/main/resources/data1.csv"
+    val logger: Logger = LoggerFactory.getLogger(this.getClass)
+    CarbonProperties.getInstance()
+      .addProperty(CarbonCommonConstants.CARBON_TIMESTAMP_FORMAT, "yyyy/MM/dd HH:mm:ss")
```
Comment (jackylk) [inline examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala]: I think default of CARBON_TIMESTAMP_FORMAT is the same, no need to set.
Can you simplify all these settings, I think they are all the same as default.

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/S3CsvExample.scala [hunk 21]
```diff
@@ -0,0 +1,113 @@
...
+import org.apache.spark.sql.SparkSession
+import org.slf4j.{Logger, LoggerFactory}
+
+import org.apache.carbondata.core.constants.CarbonCommonConstants
+import org.apache.carbondata.core.util.CarbonProperties
+
+object S3CsvExample {
+
+  /**
+   * This example demonstrate to create local store having csv on s3.
```
Comment (jackylk) [inline examples/spark2/src/main/scala/org/apache/carbondata/examples/S3CsvExample.scala]: Change to 
> This example demonstrate to create local store and load data from CSV files on S3

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala (hunk 20 above)
Comment (jackylk) [reply examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala]: Please do the same for other 3 examples also

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/MultiStoreExample.scala [hunk 22]
```diff
@@ -0,0 +1,153 @@
...
+    import org.apache.spark.sql.CarbonSession._
+    val spark = SparkSession
+      .builder()
+      .master("local")
+      .appName("CarbonSessionExample")
+      .config("spark.sql.warehouse.dir", warehouse)
+      .config("spark.driver.host", "localhost")
+      .config("spark.hadoop." + ACCESS_KEY, args(0))
+      .config("spark.hadoop." + SECRET_KEY, args(1))
+      .getOrCreateCarbonSession(storeLocation, warehouse)
```
Comment (jackylk) [inline examples/spark2/src/main/scala/org/apache/carbondata/examples/MultiStoreExample.scala]: do not give store location, use `CREATE TABLE ... LOCATION ...` to determine the location for each table

**Comment by jackylk** [comment]
Please change the merge target branch as carbonstore, since master currently has been freezed for 1.3.0 version release.

**Comment by SangeetaGulia** [comment]
@jackylk we have raised a new PR with all the review comments resolved [#1805](https://github.com/apache/carbondata/pull/1805). That PR is raised to be merged into the carbonstore branch.

Please review the same.

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/MultiStoreExample.scala (hunk 22 above)
Comment (SangeetaGulia) [reply examples/spark2/src/main/scala/org/apache/carbondata/examples/MultiStoreExample.scala]: refactored to provide store location in create table command.

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/S3CsvExample.scala (hunk 21 above)
Comment (SangeetaGulia) [reply examples/spark2/src/main/scala/org/apache/carbondata/examples/S3CsvExample.scala]: Done.

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala (hunk 20 above)
Comment (SangeetaGulia) [reply examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala]: CARBON_TIMESTAMP_DEFAULT_FORMAT = "yyyy-MM-dd HH:mm:ss" by default. So, we have not made any changes for this.

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala (hunk 19 above)
Comment (SangeetaGulia) [reply examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala]: Examples are updated with location provided in create table command.

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala (hunk 18 above)
Comment (SangeetaGulia) [reply examples/spark2/src/main/scala/org/apache/carbondata/examples/S3Example.scala]: We have taken entire location as a command line argument as the same example can be used for s3n as well.
